### DISCUSSION TITLE
Stream deck Control

### DISCUSSION THREAD

**Discussion by Hadriang** [pear-devs/pear-desktop#1650] [Ideas]
I dont even know if it's even possible but it's a feature i very much enjoyed on the other YouTube music desktop app. Wish we could have it on this much better and not outdated app x)

I can work around it using the shortcut plugin but it kinda limited ¯\_(ツ)_/¯

**Comment by ArjixWasTaken** [comment]
*For context the StreamDeck integration on [ytmdesktop](https://github.com/ytmdesktop/ytmdesktop) is [a third-party plugin](https://github.com/XeroxDev/YTMD-StreamDeck).*

I'm pinging the dev/maintainer (of said plugin) for further input.

---

Hi @XeroxDev!
Since (I assume) none of the devs/maintainers here own a StreamDeck, I'd like to request your assistance.
Can you enlighten us on how your plugin works?

Like, break down the basics (please :pray:).
I am not looking to do a direct port of your plugin, so I'd really like to learn from you instead!

**Reply by XeroxDev** [reply]
Hey @ArjixWasTaken

Thanks for pinging!

Of course I am ready and willing to help you, the users, the other contributors and repository owners! :D

Even if, due to the relatively late time here (10 pm), it's not very much now (tomorrow there will be more and in more detail), here is my feedback, confirmation and a few small pieces of information.  

For the communication with the StreamDeck I'm using my own, open source, "SDK" you can find here
https://github.com/XeroxDev/Stream-Deck-TS-SDK

For the communication with YTMDesktop I'm using the provided rest and socket io endpoints.

If I'm not wrong, this is a fork of YTMDesktop? If that's the case and you didn't drastically changed or removed said rest and socket endpoints, there's a high possibility the plugin already supports this version.

Should your api be different or doesn't work well / fully, I would suggest / offer a cooperation, so we can add support for this version to the original plugin **together**. This would have the benefit of having only one plugin in the marketplace, to support both versions, instead of multiple, which is better for the user or otherwise it could confuse some users (and probably results in user opening issues for the wrong plugin / in the wrong repository. Eg. Your users opening tickets why it's not working in my repository and vice verca.)

I could also offer, after the StreamDeck version is done of course, to implement it in my Loupedeck version / plugin as well (see https://github.com/XeroxDev/Loupedeck-plugin-YTMDesktop ).

Of course, all of this is just an offer. And if you'd like to do your own plugin, I would still help you and you can, of course, use the sdk and available plugin code as stated in their licenses (that's why I chose to make them open source, so everyone can benefit from it).

Just let me know your decision and how you all (if any other user, contributor or owner want to jump in as well, please let me know :D) / we together want to proceed in the future.

Edit: My English is terrible right now ._. My brain capacity is at 0 lol. I should really go to bed :D

**Reply by ArjixWasTaken** [reply]
Thanks for replying!

No this is not a fork of ytmdesktop!
And thanks for the tips! (and the sdk library you made)

Sadly your plugin cannot support both projects (in the same codebase) due to the way our plugin system works.

Plugins are first-party here, they are located directly in the source code of our app.

(This also means that one cannot load a third party plugin w/o rebuilding the app)

It is late for me as well so I'll converse more tomorrow!

**Reply by XeroxDev** [reply]
So... If I understood correctly, you don't offer any api or way to control your youtube music from outside the app? Then you might have a problem tbh, because an elgato plugin has to be installed and run in the elgato software itself and the plugin must have a way to communicate with you application (or how should it control or display things? 😅)

**Reply by ArjixWasTaken** [reply]
Naruhodo!

Currently there is no remote control API, but there is an upcoming PR to add that capability via a REST API

**Reply by XeroxDev** [reply]
I would also recommend for said PR, that you add a way so the server (your app) sends updates to the plugin. This would reduce the amount of requests (especially if you want a semi-live update of time etc, you'd have to send a request every second, which could be devastating for network, plugin and server performance)

**Reply by ArjixWasTaken** [reply]
Are you suggesting support for a web socket?

**Reply by XeroxDev** [reply]
Definitely! This would (I'm speaking out of experience) improve plugin, app and network performance. You could also use something like server sent event instead. As long as the message comes proactively from the server (instead of the client), this is the best option, as the server knows directly whether something has changed or not. The client would have to query it regularly, which is unnecessary in scenarios such as a paused song (because nothing changes when it is paused. However, the client does not know this, only the server does and can then start sending events again as soon as it happens doing something again)

**Reply by JellyBrick** [reply]
#1909

**Reply by Made0fGlass** [reply]
Hey everyone, I was looking through the repo to see if this had already been addressed, since I’m an active PEAR user and generally prefer it over YTMusic Desktop.

@XeroxDev as of now, PEAR includes a built-in API server plugin (currently in beta). From what I can tell, your plugin is able to reach the API, but the connection process fails and results in the following error: `Cannot read properties of undefined (reading 'code')`

My assumption is that the plugin is expecting an authorization flow that returns a code, but PEAR’s implementation may differ or not return that value in the expected format. I’m happy to help reproduce the issue or provide additional logs if needed.

@ArjixWasTaken not sure if this is something that can be addressed on your end, but it does seem like most of the required pieces are already in place now. Would love to have this working for my stream.

**Comment by Invalidm** [comment]
Does the API feature support the already existing youtube music connector for the stream deck or is that something entirely different? I am not understanding how to set it up
